syntax = "proto3";

package maarpc;

import "types.proto";
import "custom.recognizer.proto";
import "custom.action.proto";

message InstancePostTaskRequest {
  optional string handle = 1;
  optional string task = 2;
  optional string param = 3;
}

message InstanceSetTaskParamRequest {
  optional string handle = 1;
  optional string id = 2;
  optional string param = 3;
}

service Instance {
  rpc create(IdRequest) returns (HandleResponse);
  rpc destroy(HandleRequest) returns (EmptyResponse);

  rpc register_custom_recognizer(CustomRecognizerRequest)
      returns (stream CustomRecognizerResponse);
  rpc submit_custom_recognizer(SubmitCustomRecognizerRequest)
      returns (EmptyResponse);
  rpc unregister_custom_recognizer(HandleStringRequest) returns (EmptyResponse);
  rpc clear_custom_recognizer(HandleRequest) returns (EmptyResponse);

  rpc register_custom_action(CustomActionRequest)
      returns (stream CustomActionResponse);
  rpc submit_custom_action(SubmitCustomActionRequest) returns (EmptyResponse);
  rpc unregister_custom_action(HandleStringRequest) returns (EmptyResponse);
  rpc clear_custom_action(HandleRequest) returns (EmptyResponse);

  rpc bind_resource(HandleRequest) returns (EmptyResponse);
  rpc bind_controller(HandleRequest) returns (EmptyResponse);
  rpc inited(HandleRequest) returns (BoolResponse);

  rpc post_task(InstancePostTaskRequest) returns (IdResponse);
  rpc set_task_param(InstanceSetTaskParamRequest) returns (EmptyResponse);

  rpc status(HandleIdRequest) returns (StatusResponse);
  rpc wait(HandleIdRequest) returns (EmptyResponse);
  rpc all_finished(HandleRequest) returns (BoolResponse);

  rpc stop(HandleRequest) returns (EmptyResponse);

  rpc resource(HandleRequest) returns (HandleRequest);
  rpc controller(HandleRequest) returns (HandleRequest);
}
